### PR TITLE
*: refactor agent and add config controller

### PR DESCRIPTION
**Pull request by hexfusion**
Please do not review yet lots of things changing.

Built on #88

### CODE REVIEW

File: internal/agent/configcontroller/enrollment.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	klog.Infof("%swaiting for enrollment to be approved", c.logPrefix)
+	err := wait.ExponentialBackoff(c.enrollmentVerifyBackoff, func() (bool, error) {
+		return c.verifyDeviceEnrollment(ctx, device)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to verify device enrollment: %w", err)
+	}
+	if err := c.writeDeviceEnrollmentBanner(ctx, device); err != nil {
```
Comment (fzdarsky) [inline internal/agent/configcontroller/enrollment.go:34]: The banner is what allows enrolment requests to be eventually approved, so needs to happen before waiting for approval.

File: internal/agent/configcontroller/config_controller.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+	var conditions []v1alpha1.DeviceCondition
+	// post enrollment update status
+	deviceCondition := v1alpha1.DeviceCondition{
+		Type:   "Enrolled",
+		Status: v1alpha1.True,
+	}
+	conditions = append(conditions, deviceCondition)
+	deviceStatus.Conditions = &conditions
+	_, updateErr := c.managementClient.UpdateDeviceStatus(ctx, *device.Metadata.Name, deviceStatus)
```
Comment (fzdarsky) [inline internal/agent/configcontroller/config_controller.go:122]: Is this message sent on every call to sync(), even if the device is already enrolled?

File: internal/agent/configcontroller/config_controller.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+	if updateErr != nil {
+		klog.Errorf("%sfailed to update device status: %v", c.logPrefix, updateErr)
+	}
+
+	// ensure the device is configured
+	if err := c.ensureConfig(ctx, device); err != nil {
+		klog.Errorf("%s configuration did not succeed: %v", c.logPrefix, err)
```
Comment (fzdarsky) [inline internal/agent/configcontroller/config_controller.go:129]: If the configuration din't complete for whatever reason, now do we now completely roll back any changes (in a transaction with the OS image update)?

File: internal/agent/configcontroller/enrollment.go (hunk 1 above)
Comment (fzdarsky) [inline internal/agent/configcontroller/enrollment.go:34]: This needs to happen before waiting for approval, as it's pre-requisite for the approval process.

**Comment by fzdarsky** [comment]
Overall, this refactoring looks great!

One question I still have is why introduce parallelism that makes it harder to reason about the agent to ensure determinism and safety? We're like polling for status update at most 1/min while taking ms to complete, so it's not clear why we need a worker queue and async exporters/observers?

File: internal/agent/configcontroller/config_controller.go (hunk 3 above)
Comment (hexfusion) [reply internal/agent/configcontroller/config_controller.go:129]: Correct rollback is not yet supported.

File: internal/agent/configcontroller/config_controller.go (hunk 2 above)
Comment (hexfusion) [reply internal/agent/configcontroller/config_controller.go:114]: makes sense to batch this.

File: internal/agent/configcontroller/config_controller.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+	Raw  json.RawMessage `json:"inline"`
+	Name string          `json:"name"`
+}
+
+func (c *ConfigController) ensureConfig(_ context.Context, device *v1alpha1.Device) error {
+	if device.Spec.Config == nil {
+		return fmt.Errorf("device config is nil")
+	}
+
+	for _, config := range *device.Spec.Config {
```
[the remark this answers is not in the document]
Comment (hexfusion) [reply internal/agent/configcontroller/config_controller.go:191]: Right I am going off the example provided WRT inline config so we can adjust this once that flow gets nailed down.

**Comment by hexfusion** [comment]
I need to rebase

**Pull request closed without merging**